src/Build/Graph/ProjectGraph.cs

```diff
@@ -203,31 +224,54 @@ public ProjectGraph(string entryProjectFile)
 
         private ProjectGraphNode CreateNewNode(
             ConfigurationMetadata configurationMetadata,
-            ProjectCollection projectCollection)
+            ProjectCollection projectCollection,
+            bool exposeProject)
         {
             // TODO: ProjectInstance just converts the dictionary back to a PropertyDictionary, so find a way to directly provide it.
             var globalProperties = configurationMetadata.GlobalProperties.ToDictionary();
-            var project = new ProjectInstance(
-                configurationMetadata.ProjectFullPath,
-                globalProperties,
-                configurationMetadata.ToolsVersion,
-                projectCollection);
+
+            Project project = null;
+            ProjectInstance projectInstance;
+            if (exposeProject)
+            {
+                project = new Project(
+                    configurationMetadata.ProjectFullPath,
+                    globalProperties,
+                    configurationMetadata.ToolsVersion,
+                    projectCollection);
+                projectInstance = project.CreateProjectInstance(ProjectInstanceSettings.ImmutableWithFastItemLookup);
```

Comment:
I'm not super sure what `ImmutableWithFastItemLookup` does, but it seems like either we should use it in both cases, or we shouldn't use it at all (if it breaks the ability to actually build the project). MSBuild.exe doesn't use this so I suspect it breaks something.

Reply:
This setting is used in the internal build engine when flattening the Project to a ProjectInstance, for performance reasons. It has two effects in ProjectInstance.cs:

            var keepEvaluationCache = (settings & ProjectInstanceSettings.ImmutableWithFastItemLookup) == ProjectInstanceSettings.ImmutableWithFastItemLookup;

That flag leads to CreateEvaluatedIncludeSnapshotIfRequested() setting _itemsByEvaluatedInclude, which speeds up lookups in GetItemsByItemTypeAndEvaluatedInclude().

I'd prefer to stick with what is known to work in the internal build engine, but wouldn't mind having this as an additional flag in the proposed option flags noted above.

Reply:
My fear is if the flag basically makes it "parse-only, no target execution", which would be pretty broken in the general case.

The evaluation cache is probably (?) ok. The immutable thing is what I'm mostly afraid of. For example, it seems to imply to me that any property or item being set during target execution would just fail.

Someone more familiar with this setting might need to jump in and explain the details of how it effects target execution.

Reply:
It'd be the normal case by default and opt-in to this behavior with a flag, which would be used in the internal build engine pretty much exclusively AFAICT.

Reply:
If this *were* something we were building specifically for QuickBuild, it would not be in the MSBuild repo.

Reply:
It is in the public MSBuild API but IIRC was prompted by performance needs of QuickBuild.